flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -535,41 +545,82 @@ protected void beforeInvoke() throws Exception {
+                                    getEnvironment()
+                                            .getTaskStateManager()
+                                            .getSequentialChannelStateReader();
+                            reader.readOutputData(
+                                    getEnvironment().getAllWriters(),
+                                    !configuration.isGraphContainingLoops());
+
+                            operatorChain.initializeStateAndOpenOperators(
+                                    createStreamTaskStateInitializer());
+
+                            IndexedInputGate[] inputGates = getEnvironment().getAllInputGates();
+                            channelIOExecutor.execute(
+                                    () -> {
+                                        try {
+                                            reader.readInputData(inputGates);
+                                        } catch (Exception e) {
+                                            asyncExceptionHandler.handleAsyncException(
+                                                    "Unable to read channel state", e);
+                                        }
+                                    });
+
+                            List<CompletableFuture<?>> recoveredFutures =
+                                    new ArrayList<>(inputGates.length);
+                            for (InputGate inputGate : inputGates) {
+                                recoveredFutures.add(inputGate.getStateConsumedFuture());
+
+                                inputGate
+                                        .getStateConsumedFuture()
+                                        .thenRun(
+                                                () ->
+                                                        mainMailboxExecutor.execute(
+                                                                inputGate::requestPartitions,
+                                                                "Input gate request partitions"));
+                            }
 
-                    for (InputGate inputGate : getEnvironment().getAllInputGates()) {
-                        inputGate
-                                .getStateConsumedFuture()
-                                .thenRun(
-                                        () ->
-                                                mainMailboxExecutor.execute(
-                                                        inputGate::requestPartitions,
-                                                        "Input gate request partitions"));
-                    }
-                });
+                            return gatesRecoveredFuture(suspendableLoop, recoveredFutures);
+                        });
+
+        // Run mailbox until all gates will be recovered.
+        do {
+            suspendableLoop.run();
+        } while (isMailboxLoopRunning() && !allGatesRecoveredFuture.isDone());
```

Comment:
Wouldn't it be simpler to just use
```
do {
  mailboxProcessor.runMailboxStep();
} while (isMailboxLoopRunning() && !allGatesRecoveredFuture.isDone())
```
?

Reply:
I actually agree that we can do something like that:
```
do {
  mailboxProcessor.runMailboxStep(singleStep = true);
} while (isMailboxLoopRunning() && !allGatesRecoveredFuture.isDone())
```
But I am not fully sure about **singleStep = false**. Doesn't it lead to potentially infinity loop in processMailsWhenDefaultActionUnavailable if for example it doesn't have default action until state is not running but state is not running because we wait for at least one default action after allGatesRecoveredFuture was set to true? Or doesn't my example make sense?